### ISSUE TITLE
Multiple channels on one socket

### ISSUE THREAD

**Issue by ramon-janssen** [TorXakis/TorXakis#146]
Many models are most natural if they have multiple input channels, now only one is allowed (1 channel per socket). Work-arounds which currently work, have a 'unifiying' datatype to unify actual modelled data types A, B, C, ...
Unified ::= Atype{a :: A} | Btype{b :: B} | Ctype{c :: C} | ...
We can then use it in two ways:
- Give all channels in the model the same type. In every PROCDEF we need to explicitly unpack this unified datatype to the modelled data type.
- Create an extra demultiplexing process which does nothing more than receiving such a unified datatype, and (synchronously) outputting the corresponding modelled data type. The model can then have multiple channels as you would like, which are hidden and are put in parallel with this demultiplexer.

In both cases, a lot of extra boilerplate code is needed in a TorXakis model, and the model becomes less readable and unexplainable.

**Comment by pjljvandelaar** [comment]
Understand I it correctly that you purposes that TorXakis generates the Unified datatype and the boilerplate code automatically when multiple channels are mapped onto one socket?

For output I can understand this, but what about inputs: does the user know how to encode the input?

**Comment by ramon-janssen** [comment]
Inputs (for the SUT, outputs for TorXakis) could be encoded per channel: in CNECTDEF, you can then list them individually, and map them to strings individually. For Outputs (for the SUT, inputs for TorXakis) this indeed seems more difficult, at least for sockets. This could be solved with a higher-level encoding than raw strings, e.g. XML, in which you could simply specify for which channel an output action is intended. 

If I'm not mistaken, this was in an old version of TorXakis already. I guess it did not do this by generating boilerplate code but by natively mapping channels to sockets in an n-to-1 manner, which should be more efficient than generating the code as I described.

**Comment by capitanbatata** [comment]
I'm now helping @tretmans with the restart SUT functionality, which requires me to touch the code that connects with the SUT (or external world), so I would like to understand this issue to see which changes are required to make this happen.

I think adding some encoding information might make sense here. 

From the TorXakis language point of view, what it would be required is some syntax that states how multiple actions are mapped onto one socket, and how one socket can be used to encode different actions. Right? Something along these lines:

```text
CHAN OUT a PORT X
CHAN OUT b PORT X
ENCODE a ? v -> ! toString(v)

CHAN IN c PORT Y
DECODE c ! fromString(v) <- c ? v
CHAN IN d PORT Y
DECODE d ! fromString(v) <- d ? v
```

This means that actions on channels `a` and `b` can be simply sent to the socket open on port `X`, and whatever comes from the socket on port `Y` will be decoded to a `c` action if the message header matches what is expected as header on channel `c` (hence the `c ? v` part), or to `d` action, or to nothing if the headers of the message do not match any of the channels above.

Does that make sense?

Also the syntax could be simplified a bit by just having something like:

```text
{a, b} ? v -- X --> ! toString(v)

{c, d} ! fromString(v) <-- Y -- 
```

And maybe by introducing a protocol we can get rid of the `toX`/`fromX` functions.

**Comment by ramon-janssen** [comment]
Yes, that makes sense. But indeed, this solution does require you to have a higher-level protocol than raw strings, to encode the channels c and d. I guess that such a protocol is useful anyway if you want to reset the sut (just send a reset-header and let the adapter handle it), and it may also be useful for synchronous actions.

If a nice standard such as XML is used, it should be easy to set the right headers etc. (and even to encode entire TorXakis datastructures) in an external adapter, as you can use your favorite programming language and all its libraries. Whereas solving this on the side of TorXakis is tedious and has little added value (as far as I see it), in the same way in which you don't want to write your own fromString in TorXakis right now.

If the output of the SUT does not match anything, I'd say its a non-conforming output (just as a failing fromString, I don't even now if that gives a fail or a parse-error right now), but that may be a matter of preference.

**Comment by capitanbatata** [comment]
>Yes, that makes sense. But indeed, this solution does require you to have a higher-level protocol than raw strings, to encode the channels c and d (in the external adapter). I guess that such a protocol is useful anyway if you want to reset the sut (just send a reset-header and let the adapter handle it), and it may also be useful for synchronous actions.

Indeed. However be aware that it seems that right now @tretmans might have a different idea I think, where the commands to start and stop the SUT are specified in the connect defs.

> If a nice standard such as XML is used, it should be easy to set the right headers etc. (and even to encode entire TorXakis datastructures) in an external adapter, as you can use your favorite programming language and all its libraries. Whereas solving this on the side of TorXakis is tedious and has little added value (as far as I see it), in the same way in which you don't want to write your own fromString in TorXakis right now.

I guess it depends. If we use a text based protocol we might go for something text based. I won't use XML as it does not have the advantages of a text based format (readability) and it has a lot of overhead. JSON or YAML might be better alternatives. However if we want to reduce the communication overhead a binary protocol might make more sense. But even in this case we could use standard solutions such as Google's protocol buffers.

> If the output of the SUT does not match anything, I'd say its a non-conforming output (just as a failing fromString, I don't even now if that gives a fail or a parse-error right now), but that may be a matter of preference.

I guess you and @tretmans might have a better insight there regarding what TorXakis should report on unexpected output from the SUT :wink:
